llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +138,213 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
   return V;
 }
 
+SmallVector<const Value *, 5> collectIntrinsicOperands(const Value *V) {
+  const auto *Extract = dyn_cast<ExtractValueInst>(V);
+  assert(Extract && "Not an Extract");
+  const auto *IntrinsicCall =
+      dyn_cast<CallBase>(Extract->getAggregateOperand());
+  assert(IntrinsicCall && "Not an Intrinsic");
+  Intrinsic::ID ID = IntrinsicCall->getIntrinsicID();
+  SmallVector<const Value *, 5> ArgSet;
+  unsigned InPtrIdx;
+  isAIEPtrAddIntrinsic(ID, InPtrIdx);
+  // Necessary operands to disambiguate.
+  for (unsigned Op : getAddIntrinsicOps(ID)) {
+    ArgSet.push_back(IntrinsicCall->getArgOperand(Op));
+  }
+  return ArgSet;
+}
+
+struct IntrinsicChainInfo {
+  unsigned IntrinsicCalls = 0;
+  SmallVector<const Value *, 5> IntrinsicParameters;
+
+  bool hasSameNumberOfCalls(const IntrinsicChainInfo &Other) const {
+    return IntrinsicCalls == Other.IntrinsicCalls;
+  }
+
+  bool hasSameParameters(const IntrinsicChainInfo &Other) const {
+    return IntrinsicParameters == Other.IntrinsicParameters;
+  }
+
+  void copyParametersIfEmpty(const IntrinsicChainInfo &Other) {
+    if (IntrinsicParameters.empty())
+      IntrinsicParameters = Other.IntrinsicParameters;
+  }
+};
+
+/// This function returns some tracking information related to the upward path
+/// that starts from a Value *V to the Phi node that firstly defines a pointer
+/// value to the same ambiguous memory. The returned information can be
+/// described as follows:
+/// * int: the number of times this pointer is updated through a chain of
+///   add_2d/3d intrinsics calls. If these the pointer is updated in a non
+///   uniform way, ie. using a different addressing information (offset,
+///   (increment and size) this value will be -1. We also return {0 and {}}
+///   int case we track a value that represents the external Phi incoming
+///   value.
+/// * SmallPtrSet<Value*>: represent the constant (across updates) values
+///   that are used to update the pointer across different updates.
+std::optional<IntrinsicChainInfo> trackIntrinsicChain(const Value *V) {
+  IntrinsicChainInfo ChainInfo;
+  SmallVector<Value *, 5> ArgSet;
+  if (!V->getType()->isPointerTy())
+    return std::nullopt;
+  for (unsigned Count = 0; Count < BaseObjectSearchLimit; ++Count) {
+
+    if (auto *GEP = dyn_cast<GEPOperator>(V)) {
+      V = GEP->getPointerOperand();
```

Comment:
Can we really allow non-zero geps? `ptr.add.2d(a,b,c) + offset0` might alias with  `ptr.add.2d(a,b,c) + offset1`

Reply:
Maybe add a test with some non-trivial geps?

Reply:
Good point! Now we consider save only zero-indexed geps. We have now 2 tests:

- @test_with_gep_zero
- @test_with_gep_non_zero